### ISSUE TITLE
Go interfaces in 08-wasm contract API prevent clean codec usage

### ISSUE THREAD

**Issue by damiannolan** [cosmos/ibc-go#4908]
## Summary

Interfaces in Go are behavioural and do not infer concrete types when marshalling to and from JSON. 
Using interface fields such as `exported.Height` and `exported.Path` fail to unmarshal cleanly when using `encoding/json`.

## Problem Definition

The following code fails with: `json: cannot unmarshal object into Go struct field verifyMembershipMsg.verify_membership.height of type exported.Height
`

```go
msg := types.VerifyMembershipMsg{
	Height:           clienttypes.NewHeight(0, 1),
	DelayTimePeriod:  0,
	DelayBlockPeriod: 0,
	Proof:            []byte("proof"),
	Path:             commitmenttypes.NewMerklePath("ibc"),
	Value:            []byte("value"),
}

bz, err := json.Marshal(msg)
suite.Require().NoError(err)

var newMsg types.VerifyMembershipMsg
err = json.Unmarshal(bz, &newMsg)
suite.Require().NoError(err)
```

The same behaviour can be seen in this plain Go snippet: https://go.dev/play/p/4pTpCzAnS5e
It's possible to infer the type by assigning an empty pointer of the desired type. But this approach is less than desirable.

I don't think this affects rust implementations of contracts as they are parsing raw values into their own types, however when trying to leverage Go stdlib `encoding/json` the issue arises with the `MockWasmEngine` and surfaces itself here: https://github.com/cosmos/ibc-go/blob/feat/wasm-clients/modules/light-clients/08-wasm/types/mock_engine_test.go#L186 

## Proposal

I don't have a concrete proposal at the moment. But possible options could be:
- Just break the contract APIs and ensure they are using concrete types instead of interfaces. 
- Find a different approach to routing callbacks in the `MockWasmEngine`.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

**Comment by damiannolan** [comment]
This is currently blocking #4826 as well as many other testing issues

**Comment by crodriguezvega** [comment]
Closed by #4909
